types/wlr_surface.c

```diff
@@ -310,6 +310,11 @@ static void surface_apply_damage(struct wlr_surface *surface) {
 		wlr_buffer_unref(&surface->buffer->base);
 	}
 	surface->buffer = buffer;
+
+	// wlr_surface is a buffer consumer: the buffer should only be released
+	// when another one replaces it
+	wlr_buffer_ref(&buffer->base);
+	wlr_buffer_unref(&buffer->base);
```

Comment:
Why is this code referencing and immediately unreferencing the buffer?

Reply:
This is a consequence of the naming decisions explained in the PR description ("One difference is that this PR is less explicit: both consumers and producers call wlr_buffer_unref").

`wlr_surface` is a little special: it produces `wlr_buffer` objects from buffers coming from clients, but it also allows the compositor to access `wlr_surface.buffer.texture` at any time for rendering. This means `wlr_surface` is also a consumer of the buffer (because it exposes an API to read from the buffer). This ref+unref sequence allows to transition from being a producer to being a consumer.

If the naming was more explicit, this would read less weird:

```c
wlr_buffer_lock(&buffer->base); // lock the buffer to ensure we can read from this at any time
wlr_buffer_unref(&buffer->base); // we don't need a write-only reference to the buffer anymore
```

(And we would call `wlr_buffer_unlock` when the buffer gets replaced by another one.)

Do we want to change the naming? Do you have other ideas?

Reply:
I would prefer that language, yeah.